src/script/cpp_api/s_security.cpp

```diff
@@ -503,6 +504,133 @@ bool ScriptApiSecurity::safeLoadFile(lua_State *L, const char *path, const char
 	return result;
 }
 
+std::string inline _get_mod_name_from_debug(lua_Debug *info)
+{
+	std::string mod_name;
+	std::string src = info->source;
+	std::string BIN_DIR = std::string(DIR_DELIM) + "bin" + DIR_DELIM;
+	std::string MOD_DIR = BIN_DIR + ".." + DIR_DELIM + "mods" + DIR_DELIM;
+
+	size_t found = src.find(MOD_DIR);
+	if (found != std::string::npos) {
+		// Check the path whether be faked.
+		if (src.find(BIN_DIR) == found) {
+			mod_name = src.substr(found+MOD_DIR.length());
+			found = mod_name.find(DIR_DELIM_CHAR);
+			if (found != std::string::npos) mod_name = mod_name.substr(0, found);
+		} else {
+			warningstream << "Mod security:: fake "<< MOD_DIR << " folder found in mod " << src.substr(found+MOD_DIR.length()) << std::endl;
+		}
+	}
+	return mod_name;
+}
+
+std::vector<std::string> inline _get_caller_mod_names(lua_State *L) {
+  lua_Debug info;
+  std::vector<std::string> result;
+  int level = 1;
+	std::string mod_name;
+  while (lua_getstack(L, level, &info)) {
+    FATAL_ERROR_IF(!lua_getinfo(L, "S", &info), "lua_getinfo() failed");
+    mod_name = _get_mod_name_from_debug(&info);
+    if (mod_name.empty()) break;
+    if (result.size() == 0 || result.back() != mod_name) {
+      result.push_back(mod_name);
+    }
+		level++;
+  }
+  return result;
+}
+bool inline _is_in_whitelist(const std::string &setting, const std::string &mod_name)
+{
+	std::string value = g_settings->get(setting);
+	value.erase(std::remove(value.begin(), value.end(), ' '), value.end());
+	auto mod_list = str_split(value, ',');
+	return CONTAINS(mod_list, mod_name);
+}
+
+std::string inline _get_real_caller_mod_name(lua_State *L)
+{
+	std::string result;
+	// Get gamedef from registry
+	ScriptApiBase *script = ModApiBase::getScriptApiBase(L);
+	const IGameDef *gamedef = script->getGameDef();
+	if (gamedef) {
+		std::vector<std::string> caller_mods = _get_caller_mod_names(L);
+		if (caller_mods.empty()) return result;
+		std::string executor_mod_name = caller_mods.at(0);
+		const ModSpec *executor_mod = gamedef->getModSpec(executor_mod_name);
+		if (executor_mod) {
+			bool ok = false;
+			std::string prev_mod_name;
+			std::size_t i = caller_mods.size() - 1;
+			if (i == 0) return executor_mod_name;
+			// get the first event trigger
+			while (i>0) {
+				result = caller_mods.at(i);
+				prev_mod_name = caller_mods.at(i-1);
+				const ModSpec *prev_mod = gamedef->getModSpec(prev_mod_name);
+				if (!CONTAINS(prev_mod->depends, result) && !CONTAINS(prev_mod->optdepends, result)) {
+					break;
+				}
+				if (prev_mod_name == executor_mod_name) {
+					ok = true;
+					// this means the second is the trigger, the first is caller for no more callers
+					if (i == 1) result = executor_mod_name;
+					break;
+				}
```

Comment:
this needs explanation what it does and what the goal is

Reply:
Sometimes, the initiator of the action is not a builtin module, but an event trigger from a third-party module. For example, triggered by `formspecs`. The module should be written as a dependency,